### ISSUE TITLE
`PERF102` is not detected in generators or comprehensions

### ISSUE THREAD

**Issue by dosisod** [astral-sh/ruff#6638]
`PERF102` is not detected in generators or comprehensions, only in `for` loops:

```python
d = {}

for k, _ in d.items():
    print(k)

_ = (k for k, _ in d.items())
_ = [k for k, _ in d.items()]
_ = {k for k, _ in d.items()}
_ = {k: "v" for k, _ in d.items()}
```

Running:

```
$ ruff --version
ruff 0.0.284

$ ruff x.py
x.py:3:13: PERF102 [*] When using only the keys of a dict use the `keys()` method
Found 1 error.
[*] 1 potentially fixable with the --fix option.
```

Expected: A line for the `d.items()` call in the generator and each of the `list`/`set`/`dict` comprehension.

**Comment by charliermarsh** [comment]
Good call, we should catch this.

**Comment by harupy** [comment]
Can I work on this?

**Comment by harupy** [comment]
One question: To detect the loop variable usage, can we use the same approach as the `function_uses_loop_variable` rule?

https://github.com/astral-sh/ruff/blob/7e9023b6f85c28bcfdc7fe82a507d7752e84edcf/crates/ruff/src/rules/flake8_bugbear/rules/function_uses_loop_variable.rs#L283-L284
